source/nrnmatlab.cpp

```diff
@@ -291,10 +292,155 @@ void NrnRef::set_index(double x, size_t ind) {
+    }
+    return result;
+}
+
+std::vector<double> get_y3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].y;
+    }
+    return result;
+}
+
+std::vector<double> get_z3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].z;
+    }
+    return result;
+}
+
+std::vector<double> get_arc3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].arc;
+    }
+    return result;
+}
+
+std::vector<double> get_d3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].d;
+    }
+    return result;
+}
+
+// Linearly interpolates between a and b, extrapolates if f is outside the range [0, 1].
+// Added since we don't use c++ 20 or above where we could use:
+// https://en.cppreference.com/w/cpp/numeric/lerp
+double linearly_interpolate(double a, double b, double f) {
+    return a + f * (b - a);
+}
+
+double interpolate_arrays(std::vector<double> xs, std::vector<double> ys, double v) {
+    auto xs_iter = xs.begin() + 1;
+
+    while(v > *xs_iter) {
+        xs_iter++;
+    }
+
+    auto a = ys.begin() + ((xs_iter - 1) - xs.begin());
+    auto b = ys.begin() + (xs_iter - xs.begin());
+    double f = (v - *(xs_iter - 1)) / (*xs_iter - *(xs_iter - 1));
+
+    return linearly_interpolate(*a, *b, f);
+}
+
+std::vector<double> get_segment_arc(Section* sec, double low, double high) {
+    auto result = std::vector<double>();
+
+    result.push_back(low);
+
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        double arc = sec->pt3d[i].arc;
+
+        if (arc > low && arc < high) {
+            result.push_back(arc);
+        }
+    }
+
+    result.push_back(high);
+
+    return result;
+}
+
+std::vector<double> get_section_plot_data(Section* sec, ShapePlotInterface* spi) {
+    auto result = std::vector<double>();
+
+    size_t n_segments = sec->nnode - 1;
+    double sec_length = section_length(sec);
+
+    auto arcs = get_arc3d(sec);
+    auto xs = get_x3d(sec);
+    auto ys = get_y3d(sec);
+    auto zs = get_z3d(sec);
+    auto ds = get_d3d(sec);
+
+    for (size_t i = 0; i < n_segments; i++) {
+        double x_lo = (double) i / n_segments;
+        double x_hi = (double) (i + 1) / n_segments;
+        double x = (double) (i + 0.5) / n_segments;
+        x_lo *= sec_length;
+        x_hi *= sec_length;
+
+        auto segment_arc = get_segment_arc(sec, x_lo, x_hi);
+
+        double seg_value = *nrn_rangepointer(sec, hoc_lookup(spi->varname()), x);
+
+        for (size_t j = 0; j < segment_arc.size() - 1; j++) {
+            result.push_back(interpolate_arrays(arcs, xs, segment_arc[j]));
+            result.push_back(interpolate_arrays(arcs, xs, segment_arc[j + 1]));
```

Comment:
Note: this pass's `j+1` is next pass's `j`... so we're doing interpolation for essentially all points twice... is it worth trying to avoid the redundancy? or would it impair clarity too much?

Reply:
Would look something like this:
```c++
        result.push_back(interpolate_arrays(arcs, xs, segment_arc[0]));
        result.push_back(interpolate_arrays(arcs, ys, segment_arc[0]));
        result.push_back(interpolate_arrays(arcs, zs, segment_arc[0]));
        result.push_back(interpolate_arrays(arcs, ds, segment_arc[0]));
        for (size_t j = 1; j < segment_arc.size(); j++) {
            result.push_back(result[4 * j]);
            result.push_back(interpolate_arrays(arcs, xs, segment_arc[j]));
            result.push_back(result[4 * j + 1]);
            result.push_back(interpolate_arrays(arcs, ys, segment_arc[j]));
            result.push_back(result[4 * j + 2]);
            result.push_back(interpolate_arrays(arcs, zs, segment_arc[j]));
            result.push_back(result[4 * j + 3]);
            result.push_back(interpolate_arrays(arcs, ds, segment_arc[j]));
            result.push_back(seg_value);
        }

```
Or use more memory and create the interpolated versions first
Or we do that and then combine / interleave those into the output

Reply:
Indices are off, but this is probably faster.

Reply:
- There's probaly an off-by-four issue here, should it be `result.push_back(result[4 * (j-1)])`?
- Can you just add some comments for clarity, e.g. `result.push_back(result[4 * (j-1)])  // xs for segment j-1`?
- Can you rebase on main?

Then we are ready for merge